File: cpp/quantum/src/qubit_system.cpp
// quantum.cpp
// 
// Contains functions and data structures for simulating a quantum computer.

#include "qubit_system.h" 
#include "quantum_gate.h" 
#include "util.h"
#include <cmath>
#include <stdexcept>
using std::complex; 
using std::string;
using std::length_error;
using std::runtime_error;
using std::invalid_argument;

namespace quantum_algorithm_simulator {

// Constructor just calls the initializer
QubitSystem::QubitSystem (int n, int state) {
	this->coeffs = NULL; 
	this->init(n, state); 
}


// Initializer for an n-qubit system. The coefficients are initialized such that
// the system is in the pure state given by state. 
//
// Defaults to a single-qubit system in state |0>.
//
// Examples of initializing a system in state |010>:
//   q.init(3, 2)
//   or
//   q.init(3, bin_to_int("010"))
void QubitSystem::init (int n, int state) {
	this->n = n; 
	delete [] this->coeffs; 
	this->coeffs = new complex<double>[(int) pow(2, n)]; 

	if (state >= pow(2, n)) throw invalid_argument("invalid state");
	this->collapse(state); // set system to pure state 
}

// Destructor
QubitSystem::~QubitSystem () {
	delete [] this->coeffs;
}


// Returns the number of qubits in the system
int QubitSystem::N () const { 
	return this->n; 
}


// Measures the system, returning the integer whose binary representation 
// is the sequence of measured qubits. E.g., if |0110> is measured, returns 6.
int QubitSystem::measure () {
	int state = this->get_observed_state(); 
	this->collapse(state); 
	return state;
}


// Returns the observed value resulting from measuring the ith qubit in the
// system. For example, in the system 0.5|00> + 0.5|01>, measure(2) would return
// 0 or 1, each with 50% probability. 
// Measuring the system necessarily collapses it to a linear combination of pure
// states that are compatible with the observation. 
int QubitSystem::measure (int bit_index) {
	int state = this->get_observed_state();
	int bit = get_bit(state, this->n - bit_index); 

	this->collapse(bit_index, bit); 
	return bit;
}


// Like measure(), but returns the state as a string. E.g. 1 (from measure()) or
// {0, 1} (from ameasure()) would become "01". 
string QubitSystem::smeasure () {
	int state = this->measure(); 
	return int_to_bin(state, this->n); 
}


// Returns a vector of observed values resulting when all qubits in a system are
// measured. For example, in the system 0.5|00> + 0.5|01>, measure() would 
// return {0, 0} or {0, 1}, each with 50% probability. 
// Measuring the system necessarily collapses it to the observed pure state.
int * QubitSystem::ameasure () {
	int state = this->measure(); 

	int *astate = new int[this->n];
	for (int i = 0; i < this->n; i++) { 
		astate[i] = get_bit(state, this->n - i -1);  
	}

	return astate;
}

// Alternate notation for gate.act(qubits)
void operator* (const QuantumGate &gate, QubitSystem &qubits){
	gate.act(&qubits); 
} 


// Combines two systems of qubits. 
// On return, the coefficients of the system q1 are overwritten with the tensor 
// product of the coefficients of  q1 and q2.
// The system q2 is "erased" in that it is resized to a system of size zero. 
// This represents the fact that the two systems are combined into a single 
// system of separable states, and avoids violating the no-cloning theorem.  
void operator* (QubitSystem &q1, QubitSystem &q2) {
	// Number of qubits in the product system
	int N = q1.n + q2.n; 
	// coefficients of the product system
	complex<double> *coeffs = new complex<double>[(int) pow(2, N)]; 
	int index; 

	// The coefficients of the tensor product are essentially the Kronecker 
	// product of the two vectors q1.coeffs and q2.coeffs thought of as n x 1 
	// matrices. 
	for (int i = 0; i < pow(2, q1.n); i++) {
		for (int j = 0; j < pow(2, q2.n); j++) {
			index = (int) i * pow(2, q2.n) + j;
			coeffs[index] = q1.coeffs[i] * q2.coeffs[j]; 
		}
	}

	// Move coeffs to be the coefficients of q1 
	q1.n = N; 
	delete [] q1.coeffs; 
	q1.coeffs = coeffs; 

	// Erase q2
	q2.n = 0; 
	delete [] q2.coeffs; 
	q2.coeffs = NULL; // avoids memory errors when destructor is called
}


// Copy constructor. This is private so as to not allow programs to violate
// the no-cloning theorem. 
QubitSystem::QubitSystem (const QubitSystem &orig) {
	this->n = orig.n; 
	this->coeffs = copyv(orig.coeffs, pow(2, this->n)); 
}

// Copy assignment operator. This is private so as to not allow programs to 
// violate the no-cloning theorem.
QubitSystem & QubitSystem::operator= (const QubitSystem &orig) {
	// Do nothing if this == orig
	if (this == &orig) return *this;

	// If the n's are unequal, delete and reallocate coeffs
	if (this->n != orig.n) {
		this->n = orig.n; 
		delete [] this->coeffs; 
		this->coeffs = new complex<double>[(int) pow(2, this->n)]; 
	}

	// Copy coefficients in place
	copyv(orig.coeffs, this->coeffs, (int) pow(2, this->n)); 
	
	return *this; 
}

// Collapses the system of qubits to a single pure state. It sets the 
// coefficients on the state given by state to one, and all other 
// coefficients to zero.
void QubitSystem::collapse (int state) {
	for (int i = 0; i < pow(2, this->n); i++) {
		this->coeffs[i] = 0.;
	}
	this->coeffs[state] = 1.;
}


// Collapses a system of qubits to those states where the bit given by bit_index
// has the given observed value. E.g., if q is a 2-qubit system, then calling
// q.collapse(2, 1) would collapse it down to a combination of the basis states
// |01> and |11>. 
void QubitSystem::collapse (int bit_index, int observed_value) {
	// Interval over which the values 0 and 1 alternate for the bit given by 
	// bit_index. For example, if bit_index = 2 and n = 3, then interval would
	// equal 2, because the bit is zero for 2 states, then one for 2 states, then
	// zero for 2, and then one for 2: 
	// |000> |001> |010> |011> |100> |101> |110> |111>
	int interval = pow(2, this->n - bit_index); 

	// Set the probabilities of all states where the bit given by bit_index is NOT
	// equal to observed_value to zero.
	// Iterate over the first state in each "block" of consecutive states where
	// the bit given by bit_index is NOT equal to observed_value.  
	// Note that the operation "observed_value ^ 1" maps 0 to 1 and vice versa. 
	for (int i = (observed_value ^ 1) * interval; i < this->n; i += interval * 2){
		// Then, interate over each state within this block
		for (int j = 0; j < interval; j++) {
			this->coeffs[i+j] = 0.;
		}
	}

	// Then, normalize the remaining states such that their squared coefficients
	// sum to one. 
	normalize (this->coeffs, this->n); 
}


// Randomly observes a particular state of the system, based on the probability
// magnitudes. Returns the index (between 0 and 2^n-1) of the state that is
// observed. 
int QubitSystem::get_observed_state () {
	double randval = runif(); 
	double total_prob = pow(abs(this->coeffs[0]), 2); 
	int index = 0; 
	while (total_prob < randval && index < pow(2, this->n) - 1) {
		index++; 
		total_prob += pow(abs(this->coeffs[index]), 2);
	}

	return index; 
}


// Converts a string of 0s and 1s to the integer it represents in binary form. 
// E.g. if str is the string "1010", then bin_to_int(str) would return 10. 
// This is here, rather than in util.cpp where it might seem to belong better,
// so that it can be publically visible and used to do things like
// q = QubitSystem(3, bin_to_int("011")).
int bin_to_int (string bin) {
	int n = 0; 
	int len = bin.length(); 
	int bit; 

	for (int i = 0; i < len; i++) { 
		bit = bin[len-i-1];  
		if (bit == '1') { 
			n += 1 << i; // equals 2^i
		} 
		else if (bit != '0') {
			throw invalid_argument("Argument must be string of 0s and 1s only");
		}
	}

	return n; 
} 


} 

File: python/quantum/qubit_system.py
# qubit_system.py

"""Contains the class QubitSystem for representing a system of one or more
qubits."""

from math import log, sqrt
from numpy import arange, array, cumsum, kron, nonzero, random, zeros

class QubitSystem(object):
    """Represents a system of one or more qubits.

    Attributes:
        __coeffs: The coefficients on each of the basis states. __coeffs[i] is
        the coefficient on the basis state that is the binary representation of
        the integer i. For example, a 2-qubit system
        a|00> + b|01> + c|10> + d|11> would have coeffs = [a, b, c, d].
    """

    def __init__(self, n=1, state=0):
        """Constructor for a system of zero or more qubits initialized to a pure
        state.

        Args:
            n: Number of qubits in the system. Must be a nonnegative integer.
            Default 1.
            state: Pure state that the system is initialized to. The system is
            initialized to the pure state that is the binary representation of
            the integer "state".

        Raises:
            ValueError if illogical values are passed for n or state.
        """
        if n < 0 or n % 1 != 0:
            raise ValueError('n must be a nonnegative integer')
        if state < 0 or state >= 2**n or state % 1 != 0:
            raise ValueError('state must be an integer between 0 and 2^n - 1')

        self.__coeffs = zeros(2**n)
        self.__coeffs[state] = 1.


    def __repr__(self):
        return '<qubit_system.QubitSystem object at ' + hex(id(self)) + '>'


    def __str__(self):
        return 'QubitSystem of ' + str(self.n()) + ' quantum bits'


    def copy(self):
        """You can't make a copy of a QubitSystem: this enforces the no-cloning
        theorem."""
        raise TypeError('Cannot copy qubits: violation of no-cloning theorem')


    def n(self):
        """Returns the number of qubits in the system."""
        return int(round(log(len(self.__coeffs), 2)))


    def measure(self, bit_index=None):
        """Measures the state of the qubit system.

        Args:
            bit_index: If supplied, measures the state of the qubit given by
            bit_index. (One-based indexing is used for qubits, so 1 is the first
            qubit and so on.) If omitted, measures the state of all qubits in
            the system.

        Returns:
            If bit_index is supplied, returns either 0 or 1 depending on the
            measured value of the bit_index-th qubit. If bit_index is absent,
            returns an integer whose binary representation is the values of all
            the qubits. E.g., if the system is measured to be in the state 0110,
            returns 5.

        Raises:
            ValueError if bit_index is out of range.
        """
        if bit_index is not None and (bit_index < 1 or bit_index > self.n()):
            raise ValueError('bit_index out of range')

        # Randomly observe a state, based on the probability amplitudes
        rand = random.random()
        cumprobs = cumsum(abs(self.__coeffs)**2)
        state = int(min(nonzero(cumprobs >= rand)[0])) # Note 1

        if bit_index is None:
            # collapse to the observed state
            self.__coeffs = zeros(len(self.__coeffs))
            self.__coeffs[state] = 1.
            return state
        else:
            # value of bit_index-th qubit
            bit = (state >> (self.n() - bit_index)) & 1
            # collapse only the bit_index-th qubit to the observed value
            self.__collapse_qubit(bit_index, bit)
            return bit


    def smeasure(self):
        """Measures the state of the qubit system as a binary string.

        Returns:
            A string of 0's and 1's corresponding to the observed values of each
            qubit. E.g., if the system is observed in the state 011, returns
            "011".
        """
        state = bin(self.measure())[2:] # state as binary int sans leading zeros
        return state.zfill(self.n())


    def __collapse_qubit(self, bit_index, observed_value):
        """Collapses the system of qubits to those states where the bit given by
        bit_index has the given observed value.

        E.g., if q is a two-qubit system, then calling q.__collapse_qubit(2,1)
        would collapse it to a superposition of the states |01> and |11>.

        Args:
            bit_index: Index of qubit to collapse into a particular state.
            observed_value: The value of the bit_index-th bit that was observed,
            either 0 or 1.
        """
        # Interval over which the values 0 and 1 alternate for the bit given by
        # bit_index. For example, if bit_index = 2 and n = 3, then interval
        # would equal 2, because the bit is zero for 2 states, then one for 2
        # states, then zero for 2, and then one for 2:
        # |000> |001> |010> |011> |100> |101> |110> |111>
        interval = 2**(self.n() - bit_index)

        # Set the probabilities of all states where the bit given by bit_index
        # is NOT equal to observed_value to zero.
        # Iterate over the first state in each "block" of consecutive states
        # where the bit given by bit_index is NOT equal to observed_value. Note
        # that the operation "observed_value ^ 1" maps 0 to 1 and vice versa.
        for i in range((observed_value ^ 1) * interval, self.n(), 2 * interval):
            self.__coeffs[i+arange(interval)] = 0.

        # Then, normalize the remaining states such that their squared
        # coefficients sum to one.
        self.__coeffs /= sqrt(sum(abs(self.__coeffs)**2))


    def __mul__(self, other):
        """Combines two systems of qubits.

        On return, the coefficients on the current system of qubits are replaced
        by the tensor product of the two sets of coefficients. The system
        "other" is "erased" in that it is resized to a system of size zero. This
        represents the fact that the two systems are combined into a single
        system of separable states, and avoids violating the no-cloning theorem.

        Args:
            other: Another system of qubits to join with the current system.
        """
        self.__coeffs = kron(self.__coeffs, other.__coeffs)
        other.__coeffs = array([])



# NOTES
# 1. The int() cast is to change the output of nonzero() from int64, which can
#    cause weird errors - noticed when calling pow(a, b, c) where b is an int64
#    and a, c are ints.




